design_docs/mpas_xarray_class.md

```diff
@@ -0,0 +1,185 @@
+own module so that it remains autonomous from MPAS-Analysis.
+
+At the same time, it is desirable for efficiency and parallelism to perform
+certain operations during the preprocessing step within xarray, rather than
+constructing a data set first and then (in serial) performing manipulations
+(e.g. creating a time coordinate and slicing variables).  
+
+To solve these problems, this document proposes a solution in which a new 
+MpasXarray class is created with two methods, `open_multifile_dataset` and
+`_preprocess` that can be overloaded by subclasses.  A subclass, 
+MpasAnalysisXarray, is also added in which these methods are overloaded to
+add support for (among other things) mapping variables.
+
+The solution will be tested by making sure it produces bit-for-bit identical
+results to those from the develop branch for typical test cases on LANL IC
+and Edison.
+
+<h1> Requirements </h1>
+
+<h2> Requirement: mpas_xarray does not include MPAS-Analysis specific
+functionality <br>
+Date last modified: 2017/02/10 <br>
+Contributors: Xylar Asay-Davis
+</h2>
+
+MPAS-Analysis specific functionality such as variable mapping should be 
+removed from mpas_xarray so it can remain an independent module, requiring
+minimal modification to accommodate MPAS-Analysis' needs.
+
+<h2> Requirement: MPAS-Analysis specific functionality should be supported in
+xarray preprossing <br>
+Date last modified: 2017/02/10 <br>
+Contributors: Xylar Asay-Davis
+</h2>
+
+There should be a way to perform MPAS-Analysis specific functionality such as
+mapping variables during preprocessing.  This functionality should be
+relatively easy to add to as new preprocessing needs arise.
+
+
+<h1> Algorithmic Formulations (optional) </h1>
+
+<h2> Algorithm: mpas_xarray does not include MPAS-Analysis specific
+functionality <br>
+Date last modified: 2017/02/10 <br>
+Contributors: Xylar Asay-Davis
+</h2>
+
+All functions and function arguments related to variable mapping will
+be removed from mpas_xarray and moved elsewhere.
+
+<h2> Algorithm: MPAS-Analysis specific functionality should be supported in
+xarray preprossing <br>
+Date last modified: 2017/02/10 <br>
+Contributors: Xylar Asay-Davis
+</h2>
+
+While this requirement seems almost contradictory to the previous
+requirement, both can be accommodated by using classes and inheritance.
+An MpasXarray class will be added to mpas_xarray, including a method
+(`_preprocess()`) for performing preprocessing.  Then, a subclass will
+be created (MpasAnalysisXarray) that includes the functionality that
+is specific to MPAS-Analysis by overloading the preprocessing method
+from MpasXarray.
+
+This solution may add a level of complexity, particularly for
+developing mpas_xarray and its analysis-specific derived reader.
+It also risks confusing a new developer by having seemingly redundant
+functionality in two places.  However, it seems to me to be the
+cleanest way to satisfy both requirements.
```

Comment:
Okay, I'm learning your terminology as I go.  Now that I understand what functional and class-composition approaches are, I think a functional approach could be a clear alternative.  A class-composition approach might not work without effectively combining it with a functional approach because MpasXarray needs to know what preprocess function to call.

Over all, I'm in favor of considering a purely functional approach.  I will see how far I can get with that modification.  If it works, it will probably make things simpler.  Effectively, we would be returning to mpas_xarray as it was with some clean up and an additional `open_multifile_dataset` function.